crates/libs/core/src/sync/cancel.rs

```diff
@@ -0,0 +1,729 @@
+        let prev = self.content.replace(Some(content));
+        assert!(prev.is_none())
+    }
+
+    // can only be called once after set content.
+    fn consume_content(&self) -> crate::Result<T> {
+        let opt = self.content.take();
+        match opt {
+            Some(x) => Ok(x),
+            None => {
+                if !self.IsCompleted().as_bool() {
+                    return Err(FabricErrorCode::AsyncOperationNotComplete.into());
+                }
+                if self.token.is_cancelled() {
+                    Err(FabricErrorCode::OperationCanceled.into())
+                } else {
+                    panic!("content is consumed twice.")
+                }
+            }
+        }
+    }
+
+    fn set_complete(&self) {
+        self.is_completed.swap(&Cell::new(true));
+    }
+}
+
+impl<T> IFabricAsyncOperationContext_Impl for BridgeContext3<T> {
+    fn IsCompleted(&self) -> ::windows::Win32::Foundation::BOOLEAN {
+        self.is_completed.get().into()
+    }
+
+    // This always returns false because we defer all tasks in the background executuor.
+    fn CompletedSynchronously(&self) -> ::windows::Win32::Foundation::BOOLEAN {
+        self.is_completed_synchronously.into()
+    }
+
+    fn Callback(&self) -> ::windows_core::Result<IFabricAsyncOperationCallback> {
+        let cp = self.callback.clone();
+        Ok(cp)
+    }
+
+    fn Cancel(&self) -> ::windows_core::Result<()> {
+        self.token.cancel();
+        Ok(())
+    }
+}
+
+// proxy impl
+
+// Token that wraps oneshot receiver.
+// The future recieve does not have error. This is designed for the use
+// case where SF guarantees that sender will be called.
+pub struct FabricReceiver2<T> {
+    rx: tokio::sync::oneshot::Receiver<T>,
+    token: Option<CancellationToken>,
+    // saved ctx from SF Begin COM api for cancalling.
+    ctx: Option<IFabricAsyncOperationContext>,
+}
+
+impl<T> FabricReceiver2<T> {
+    fn new(
+        rx: tokio::sync::oneshot::Receiver<T>,
+        token: Option<CancellationToken>,
+    ) -> FabricReceiver2<T> {
+        FabricReceiver2 {
+            rx,
+            token,
+            ctx: None,
+        }
+    }
+
+    // This does not handle cancel. It is commented out because it is not used.
+    // pub fn blocking_recv(self) -> crate::Result<T> {
+    //     if let Some(t) = self.token {
+    //         if t.is_cancelled() {
+    //             return Err(FabricErrorCode::OperationCanceled.into());
+    //         }
+    //     }
+    //     // sender must send stuff so that there is not error.
+    //     Ok(self.rx.blocking_recv().unwrap())
+    // }
+
+    // Set the SF ctx to hook up cancellation.
+    fn set_ctx(&mut self, ctx: IFabricAsyncOperationContext) {
+        let prev = self.ctx.replace(ctx);
+        assert!(prev.is_none());
+    }
+
+    // Cancels the inner SF operation if exists, and reset the ctx.
+    fn cancel_inner_ctx(&mut self) -> crate::Result<()> {
+        if let Some(ctx) = &self.ctx {
```

Comment:
Would we ever expect self.ctx to be None during normal/correct operation? ie. should it have a debug_assert!()?

Reply:
Added comments about this. This might happen when polling happened multiple times and we call cancelled inner ctx multiple times and the first time will empty the ctx after cancel.